### PR TITLE
Sector Nord AG: Added ProcessManagement TransitionAction - AppointmentUpdate and AppointmentRemove.

### PR DESCRIPTION
**Pull request by jross93**
## Proposed change

Currently it is only possible to create appointments from a transition action. In consultation with Shawn Beasley we added the transition actions to update and delete an appointment.

- '1 - 🚀 feature'

## Additional information

I also updated the transition action template to help users finding the correct parameters. (ProcessManagement.xml)

## Checklist

- [x] The code change is tested and works locally.(❗)
- [x] There is no commented out code in this PR.(❕)
- [ ] You improved or added new unit tests.(❕)
- [x] Local ZnunyCodePolicy run passes successfully.(❕)
- [ ] Local unit tests pass.(❕)
- [x] GitHub workflow ZnunyCodePolicy passes.(❗)
- [x] GitHub workflow unit tests pass.(❗)

### CODE REVIEW

**Review by dennykorsukewitz — COMMENTED** [review]
Hi @jsinagowitz 

I've only had a very brief look at the code, so I'll just make a few quick points before taking a closer look (after your changes).

- great that you extend the SysConfig `ProcessManagement::TransitionAction::DefaultParameters###001-Framework` 👍🏼 
- **AppointmentDelete** => **AppointmentRemove** (DynamicFieldRemove) please keep the same in the context. (I do not like 'remove' either, because the function ends with 'delete'. But we should stay consistent here.)
- I have to tell you something... I am a UnitTest fan. I know I'm annoying you contributors with this, but someone has to be the annoying one. 😈 

Thanks 🚀

**Comment by jross93** [comment]
Hi @dennykorsukewitz 

I just renamed the module AppointmentDelete and added some unittests.

Please have a look 🚀

File: Kernel/Config/Files/XML/ProcessManagement.xml [hunk 1]
```diff
@@ -766,6 +766,32 @@
...
+                        <Item Key="RecurrenceUntil">(optional) 2016-01-10 00:00:00</Item>
+                        <Item Key="NotificationTime">(optional) 2016-01-01 17:00:00</Item>
+                        <Item Key="NotificationTemplate">(optional) Custom</Item>
+                        <Item Key="NotificationCustom">(* required for Custom) relative</Item>
+                        <Item Key="NotificationCustomRelativeUnitCount">(* required for Custom) 12 (number of time units)</Item>
+                        <Item Key="NotificationCustomRelativeUnit">(* required for Custom) minutes|hours|days</Item>
+                        <Item Key="NotificationCustomRelativePointOfTime">(* required for Custom) beforestart|afterstart|beforeend|afterend (Possible notification types for the "relative" type)</Item>
+                    </Hash>
+                </Item>
+                <Item Key="AppointmentDelete">
```
Comment (dennykorsukewitz) [inline Kernel/Config/Files/XML/ProcessManagement.xml]:
```suggestion
                <Item Key="AppointmentRemove">
```

File: Kernel/System/ProcessManagement/TransitionAction/AppointmentRemove.pm [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+# did not receive this file, see http://www.gnu.org/licenses/agpl.txt.
+# --
+
+package Kernel::System::ProcessManagement::TransitionAction::AppointmentRemove;
+
+use strict;
+use warnings;
+use utf8;
+
+use Kernel::System::VariableCheck qw(:all);
```
Comment (dennykorsukewitz) [inline Kernel/System/ProcessManagement/TransitionAction/AppointmentRemove.pm]: If you are not use one of the VariableCheck functions, you can delete this use.

File: scripts/test/ProcessManagement/TransitionAction/AppointmentRemove.t [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+# did not receive this file, see http://www.gnu.org/licenses/agpl.txt.
+# --
+
+use strict;
+use warnings;
+use utf8;
+
+use vars (qw($Self));
+
+use Kernel::System::VariableCheck qw(:all);
```
Comment (dennykorsukewitz) [inline scripts/test/ProcessManagement/TransitionAction/AppointmentRemove.t]: Is this needed?

File: scripts/test/ProcessManagement/TransitionAction/AppointmentRemove.t [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+$Kernel::OM->ObjectParamAdd(
+    'Kernel::System::UnitTest::Helper' => {
+        RestoreDatabase => 1,
+    },
+);
+
+my $AppointmentObject  = $Kernel::OM->Get('Kernel::System::Calendar::Appointment');
+my $CalendarObject     = $Kernel::OM->Get('Kernel::System::Calendar');
+my $ConfigObject       = $Kernel::OM->Get('Kernel::Config');
```
Comment (dennykorsukewitz) [inline scripts/test/ProcessManagement/TransitionAction/AppointmentRemove.t]: my $ConfigObject       = $Kernel::OM->Get('Kernel::Config');
Is this needed?

File: scripts/test/ProcessManagement/TransitionAction/AppointmentRemove.t [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+my $BackendObject      = $Kernel::OM->Get('Kernel::System::DynamicField::Backend');
+
+my $TransitionActionObject = $Kernel::OM->Get('Kernel::System::ProcessManagement::TransitionAction::AppointmentRemove');
+
+my $GroupID = $GroupObject->GroupLookup(
+    Group => 'users',
+);
+
+my %Calendar = $CalendarObject->CalendarCreate(
+    CalendarName => 'Calendar 1',
```
Comment (dennykorsukewitz) [inline scripts/test/ProcessManagement/TransitionAction/AppointmentRemove.t]: If the calendar already exists, you will have a problem. 
Either you update the calendar or you use RandomID in Name etc.

File: scripts/test/ProcessManagement/TransitionAction/AppointmentRemove.t [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    GroupID      => $GroupID,
+    Color        => '#FF7700',
+    UserID       => 1,
+    ValidID      => 1,
+);
+
+my @DynamicFields = (
+    {
+        Name       => 'appid',
```
Comment (dennykorsukewitz) [inline scripts/test/ProcessManagement/TransitionAction/AppointmentRemove.t]: Please always use very meaningful variable names. 
appid = AppointmentID

File: scripts/test/ProcessManagement/TransitionAction/AppointmentRemove.t [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+my $TransitionActionResult = $TransitionActionObject->Run(
+    UserID                   => 1,
+    Ticket                   => \%Ticket,
+    ProcessEntityID          => 'P123',
+    ActivityEntityID         => 'A123',
+    TransitionEntityID       => 'T123',
+    TransitionActionEntityID => 'TA123',
+    Config                   => {
+        AppointmentID => '<OTRS_TICKET_DynamicField_appid>',
```
Comment (dennykorsukewitz) [inline scripts/test/ProcessManagement/TransitionAction/AppointmentRemove.t]: Top! I like that very much that you're testing this. 👍🏼 👍🏼

File: Kernel/System/ProcessManagement/TransitionAction/AppointmentUpdate.pm [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+        String => '2019-01-01',
+    );
+
+Returns:
+
+    my $TimeStamp = '2019-01-01 00:00:00';
+
+=cut
+
+sub _ValidDateTimeConvert {
```
Comment (dennykorsukewitz) [inline Kernel/System/ProcessManagement/TransitionAction/AppointmentUpdate.pm]: The function has now been defined twice.
In 
**Kernel::System::ProcessManagement::TransitionAction::AppointmentCreate**
and 
**Kernel::System::ProcessManagement::TransitionAction::AppointmentUpdate.**

So this should now be moved to **Kernel::System::ProcessManagement::TransitionAction::Base;**.

File: scripts/test/ProcessManagement/TransitionAction/AppointmentUpdate.t [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+$Kernel::OM->ObjectParamAdd(
+    'Kernel::System::UnitTest::Helper' => {
+        RestoreDatabase => 1,
+    },
+);
+
+my $AppointmentObject  = $Kernel::OM->Get('Kernel::System::Calendar::Appointment');
+my $CalendarObject     = $Kernel::OM->Get('Kernel::System::Calendar');
+my $ConfigObject       = $Kernel::OM->Get('Kernel::Config');
```
Comment (dennykorsukewitz) [inline scripts/test/ProcessManagement/TransitionAction/AppointmentUpdate.t]: Is this needed?
Please also note the same comments as in the UnitTest AppointmentRemove.t here

**Review by dennykorsukewitz — CHANGES_REQUESTED** [review]
Hi @jsinagowitz

thank you.
Please use always the CodePolicy to check you code.

https://github.com/znuny/ZnunyCodePolicy/blob/master/doc/en/feature.md#usage

## Usage
```
perl $WORKSPACE/ZnunyCodePolicy/bin/otrs.CodePolicy.pl --file Test.pm
perl $WORKSPACE/ZnunyCodePolicy/bin/otrs.CodePolicy.pl --file Kernel/System/ProcessManagement/TransitionAction/AppointmentRemove.pm
```
 
## Result
```bash
 TidyAll::Plugin::Znuny::Perl::ObjectDependencies

The following objects are not longer used in the code, but declared as dependencies:

    'Kernel::System::DateTime',


[tidied]  Kernel/System/ProcessManagement/TransitionAction/AppointmentRemove.pm
```
`Kernel::System::DateTime` is obsolete... etc.


--- 

Please have a look at my comments, start the CodePolicy and improve the code accordingly.

I will certainly notice something else in the next QA round, but I am already very satisfied for now.

All in all: good job, so far. :rocket:
Thanks

**Comment by jross93** [comment]
Hi @dennykorsukewitz 

I added all improvements to the code. I shouldn't have trusted the AppointmentCreate.t. 😄 
I used it for the base. So I just updated the changes to this unit test too!

Wishing you a nice weekend 🚀

**Review by dennykorsukewitz — APPROVED** [review]

**Pull request merged**